erts/emulator/beam/jit/arm/instr_bs.cpp

```diff
@@ -1443,8 +1482,174 @@ struct BscSegment {
+            }
+
+            auto prev = segs.back();
+            if (prev.flags & BSF_LITTLE) {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+                continue;
+            }
+
+            if (prev.effectiveSize + seg.effectiveSize <= 64) {
+                segs.pop_back();
+                prev.effectiveSize += seg.effectiveSize;
+                seg.action = BscSegment::ACCUMULATE;
+                segs.push_back(seg);
+                segs.push_back(prev);
+            } else {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+            }
+            break;
+        }
+        default:
+            segs.push_back(seg);
+            break;
+        }
+    }
+    return segs;
+}
+
+void BeamModuleAssembler::update_bin_state(arm::Gp bin_base,
+                                           arm::Gp bin_offset,
+                                           Sint bit_offset,
+                                           Sint size,
+                                           arm::Gp size_reg) {
+    int cur_bin_offset =
+            offsetof(ErtsSchedulerRegisters, aux_regs.d.erl_bits_state) +
+            offsetof(struct erl_bits_state, erts_current_bin_);
+    arm::Mem mem_bin_base = arm::Mem(scheduler_registers, cur_bin_offset);
+    arm::Mem mem_bin_offset =
+            arm::Mem(scheduler_registers, cur_bin_offset + sizeof(Eterm));
+
+    if (bit_offset % 8) {
+        /* The bit offset is unknown or not byte-aligned. */
+        ERTS_CT_ASSERT_FIELD_PAIR(struct erl_bits_state,
+                                  erts_current_bin_,
+                                  erts_bin_offset_);
+        a.ldp(bin_base, bin_offset, mem_bin_base);
+
+        if (size_reg.isValid()) {
+            a.add(TMP1, bin_offset, size_reg);
+        } else {
+            a.add(TMP1, bin_offset, imm(size));
+        }
+        a.str(TMP1, mem_bin_offset);
+
+        a.add(TMP1, bin_base, bin_offset, arm::lsr(3));
+    } else {
+        comment("optimized updating of binary construction state");
+        ASSERT(size >= 0);
+        ASSERT(bit_offset % 8 == 0);
+        a.ldr(TMP1, mem_bin_base);
+        mov_imm(TMP2, bit_offset + size);
+        a.str(TMP2, mem_bin_offset);
+        add(TMP1, TMP1, bit_offset >> 3);
+    }
+}
+
+/*
+ * The size of the segment is assumed to be in ARG3.
+ */
+void BeamModuleAssembler::set_zero(Sint effectiveSize) {
+    Label words = a.newLabel();
+    Label less_than_a_word = a.newLabel();
+    Sint store_unit = 1;
+
+    update_bin_state(ARG1, ARG2, -1, -1, ARG3);
+
+    if (effectiveSize >= 128) {
+        /* Store two 64-bit words machine words when the size is
+         * known and at least 128 bits. */
+        store_unit = 2;
+    }
```

Comment:
As an aside, we can use a `store_unit` of 4 by using `STP` with vector registers.

Reply:
OK. Done that.